index.js

```diff
@@ -0,0 +1,180 @@
+const async = require('async');
+
+class J5sExecutor extends Executor {
+
+    /**
+     * Constructor
+     * @method constructor
+     * @param  {Object} options           Configuration options
+     * @param  {Object} options.host      Jenkins hostname to make requests to
+     * @param  {Object} options.username  Jenkins username
+     * @param  {Object} options.password  Jenkins password/token
+     */
+    constructor(options) {
+        super();
+        this.host = options.host;
+        this.username = options.username;
+        this.password = options.password;
+        // need to pass port nubmer in the future
+        this.baseUrl = `http://${this.username}:${this.password}@${this.host}:8080`;
+        this.crumbUrl = this.baseUrl.concat('/crumbIssuer/api/json');
+    }
+
+    /**
+     * Get CSRF token crumb for Jenkins
+     * @method getCrumb
+     * @param  {Function} callback          Callback with error and crumb object
+     */
+    getCrumb(callback) {
+        const options = {
+            url: this.crumbUrl,
+            method: 'GET'
+        };
+
+        request(options, (error, response) => {
+            if (error) return callback(new Error(error.message));
+
+            if (response.statusCode !== 200) {
+                const msg = `Failed to get crumb: ${JSON.stringify(response.body)}`;
+
+                return callback(new Error(msg));
+            }
+
+            return callback(null, response.body);
+        });
+    }
+
+    /**
+     * Initialize the jenkins client with crumb
+     * @method initJenkinsClient
+     * @param  crumb                        CSRF token from jenkins api
+     * @param  {Function} callback          Callback with error and jenkins client object
+     */
+    initJenkinsClient(crumb, callback) {
+        const data = JSON.parse(crumb);
+
+        const jenkinsClient = jenkins({
+            baseUrl: this.baseUrl,
+            headers: {
+                [data.crumbRequestField]: data.crumb
+            }
+        });
+
+        if (jenkinsClient) return callback(null, jenkinsClient);
+
+        return callback(new Error('Failed to instantiate jenkins client'));
+    }
+
+    /**
+     * Create a jenkins job and start the build
+     * @method _start
+     * @param  {Object}   config            A configuration object
+     * @param  {String}   config.buildId    ID for the build and also name of the job in jenkins
+     * @param  {String}   config.jobId      ID for the job
+     * @param  {String}   config.pipelineId ID for the pipeline
+     * @param  {String}   config.container  Container for the build to run in
+     * @param  {String}   config.scmUrl     Scm URL to use in the build
+     * @param  {Function} callback          Callback with null if successful otherwise error
+     */
+    _start(config, callback) {
+        const fakeConfigPath = path.resolve(__dirname, './config/test-job.xml');
+        const xml = fs.readFileSync(fakeConfigPath, 'utf-8');
```

Comment:
You already have a controlled work flow below. Do we want this to be a blocking operation?